File: frontend/src/types/note.ts
/** Mirrors backend/app/schemas/note.py. */

export type NoteVisibility = "PRIVATE" | "SHARED";

export interface NoteResponse {
  id: string;
  application_id: string | null;
  candidate_id: string | null;
  job_id: string | null;
  author_id: string;
  author_name: string;
  title: string | null;
  body: string;
  category: string | null;
  color: string | null;
  visibility: NoteVisibility;
  created_at: string;
  updated_at: string;
}

export interface StandaloneNoteCreateRequest {
  body: string;
  title?: string | null;
  category?: string | null;
  color?: string | null;
  visibility?: NoteVisibility;
  application_id?: string | null;
  candidate_id?: string | null;
  job_id?: string | null;
}

export interface NoteUpdateRequest {
  body?: string;
  title?: string | null;
  category?: string | null;
  color?: string | null;
  visibility?: NoteVisibility;
  application_id?: string | null;
  candidate_id?: string | null;
  job_id?: string | null;
}

export interface NoteListQuery {
  search?: string;
  category?: string;
  visibility?: NoteVisibility;
  candidateId?: string;
  jobId?: string;
  sort?: "created_at" | "updated_at" | "title";
  order?: "asc" | "desc";
}

File: backend/app/schemas/note.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.note import NoteVisibility

MAX_BODY_CHARS = 5000


class NoteCreateRequest(BaseModel):
    """Used by the application-scoped endpoint
    (`/applications/{id}/notes`) — `application_id` comes from the URL
    path, not the body. Defaults to SHARED, preserving exactly what every
    note on this endpoint already did before visibility existed."""

    body: str = Field(min_length=1, max_length=MAX_BODY_CHARS)
    title: str | None = Field(default=None, max_length=255)
    category: str | None = Field(default=None, max_length=100)
    color: str | None = Field(default=None, max_length=30)
    visibility: NoteVisibility = NoteVisibility.SHARED


class StandaloneNoteCreateRequest(BaseModel):
    """Used by the standalone Notes workspace (`/recruiter/notes`) — every
    link is optional and explicit. Defaults to PRIVATE: a personal note a
    recruiter jots down is theirs alone unless they choose to share it."""

    body: str = Field(min_length=1, max_length=MAX_BODY_CHARS)
    title: str | None = Field(default=None, max_length=255)
    category: str | None = Field(default=None, max_length=100)
    color: str | None = Field(default=None, max_length=30)
    visibility: NoteVisibility = NoteVisibility.PRIVATE
    application_id: uuid.UUID | None = None
    candidate_id: uuid.UUID | None = None
    job_id: uuid.UUID | None = None


class NoteUpdateRequest(BaseModel):
    """Every field is optional; a field left out of the request body is
    left unchanged (the route layer uses `model_fields_set` to tell
    "omitted" from "explicitly set to null" for the nullable link fields)."""

    body: str | None = Field(default=None, min_length=1, max_length=MAX_BODY_CHARS)
    title: str | None = Field(default=None, max_length=255)
    category: str | None = Field(default=None, max_length=100)
    color: str | None = Field(default=None, max_length=30)
    visibility: NoteVisibility | None = None
    application_id: uuid.UUID | None = None
    candidate_id: uuid.UUID | None = None
    job_id: uuid.UUID | None = None


class NoteResponse(BaseModel):
    id: uuid.UUID
    application_id: uuid.UUID | None
    candidate_id: uuid.UUID | None
    job_id: uuid.UUID | None
    author_id: uuid.UUID
    author_name: str
    title: str | None
    body: str
    category: str | None
    color: str | None
    visibility: NoteVisibility
    created_at: datetime
    updated_at: datetime
